atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/RangeOfTimestampsIterator.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra;
+
+import java.util.Arrays;
+import java.util.Set;
+import java.util.SortedMap;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Maps;
+import com.google.common.collect.PeekingIterator;
+import com.google.common.primitives.UnsignedBytes;
+import com.palantir.atlasdb.keyvalue.api.CandidateCellForSweeping;
+import com.palantir.atlasdb.keyvalue.api.CandidateCellForSweepingRequest;
+import com.palantir.atlasdb.keyvalue.api.ImmutableCandidateCellForSweepingRequest;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.RangeRequest;
+import com.palantir.atlasdb.keyvalue.api.RowResult;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.common.base.ClosableIterator;
+
+public class RangeOfTimestampsIterator implements ClosableIterator<RowResult<Set<Long>>> {
+    private final ClosableIterator<CandidateCellForSweeping> delegate;
+    private final PeekingIterator<CandidateCellForSweeping> candidates;
+    private final RangeRequest rangeRequest;
+
+    public RangeOfTimestampsIterator(KeyValueService keyValueService, TableReference tableRef,
+            RangeRequest rangeRequest, long timestamp) {
+        CandidateCellForSweepingRequest request = ImmutableCandidateCellForSweepingRequest.builder()
+                .startRowInclusive(rangeRequest.getStartInclusive())
+                .maxTimestampExclusive(timestamp)
+                .addTimestampsToIgnore()
+                .shouldCheckIfLatestValueIsEmpty(false)
+                .build();
+        this.rangeRequest = rangeRequest;
+        this.delegate = keyValueService.getCandidateCellsForSweeping(tableRef, request)
```

Comment:
This is pretty odd. Inside the constructor, we perform a hefty-looking KVS call. Not only that, but this is the place where we passed in `this` earlier. Is this an attempt to defer some computation?

Reply:
no, we could just pass in the delegate iterator directly, but i wanted to avoid having this logic in the KVS. could move this to a static constructor?

Reply:
A static constructor sounds OK, if this class survives any redesign.